File: sdks/csharp/auth/src/OuraToolkit.Auth/TokenManager.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace OuraToolkit.Auth;

/// <summary>
/// Owns the current tokens and the machinery to keep them fresh: proactive refresh inside a
/// skew window before expiry, refresh-token ROTATION persistence (Oura invalidates the
/// previous refresh token on every refresh), and the cross-process reload/adopt/retry
/// protocol shared with the Rust companion.
///
/// This library is auth-plumbing only — no browser, no loopback listener, no interactive
/// consent (that is CLI territory). Wire it into the generated client's bearer seam:
/// <code>
/// var manager = TokenManager.Load();
/// var config = new OuraToolkit.Api.Client.Configuration
/// {
///     AccessToken = await manager.GetAccessTokenAsync(), // fresh: refreshed + persisted if needed
/// };
/// var sleep = new OuraToolkit.Api.Api.DailySleepRoutesApi(config);
/// </code>
/// Re-call <see cref="GetAccessTokenAsync"/> (and re-assign <c>Configuration.AccessToken</c>)
/// before each request batch so long-lived processes pick up refreshes; on a 401 from the
/// data plane, call <see cref="ForceRefreshAsync"/> and retry the request once.
/// </summary>
public sealed class TokenManager : IDisposable
{
    /// <summary>Refresh this many seconds before the token's actual expiry.</summary>
    public const long DefaultSkewSeconds = 60;

    /// <summary>
    /// Hard timeout on each token-endpoint call. Load-bearing: the refresh runs under the
    /// store's exclusive lock, so this bounds how long one process's stalled refresh can
    /// wedge every other process waiting on the lock (worst case ~2×: the 400-retry arm can
    /// chain a second endpoint call under the same lock).
    /// </summary>
    public static readonly TimeSpan TokenEndpointTimeout = TimeSpan.FromSeconds(30);

    private readonly TokenStore _store;
    private readonly ClientCredentials? _credentials;
    private readonly HttpClient _http;
    private readonly string _tokenUrl;
    private readonly long _skewSeconds;
    private readonly SemaphoreSlim _mutex = new(1, 1);
    private Tokens? _tokens;

    /// <summary>
    /// Load from the default token store. Absent records are not an error here —
    /// <see cref="GetAccessTokenAsync"/> reports <see cref="NotAuthenticatedException"/> on
    /// first use, so callers can surface their own "run oura auth login" UX.
    /// </summary>
    public static TokenManager Load()
    {
        var store = new TokenStore();
        return new TokenManager(store, store.LoadCredentials(), store.LoadTokens());
    }

    /// <summary>
    /// Construct from an explicit store + optional in-memory records. Both records are
    /// independently optional: credentials-without-tokens is "setup done, no login yet";
    /// tokens-without-credentials is a caller-supplied token, usable until expiry but not
    /// refreshable (<see cref="MissingClientCredentialsException"/>).
    /// </summary>
    /// <param name="store">The on-disk store refreshes reload from and persist to.</param>
    /// <param name="credentials">The user's own OAuth app credentials (null: refresh impossible).</param>
    /// <param name="tokens">The starting token set (null: not authenticated yet).</param>
    /// <param name="handler">
    /// Token-endpoint transport override (hermetic tests inject a mock
    /// <see cref="HttpMessageHandler"/>). Default: a plain handler — deliberately NOT an
    /// authenticated client, to avoid refresh recursion.
    /// </param>
    /// <param name="tokenUrl">
    /// Token-endpoint override (tests point at a mock). Default:
    /// <see cref="OAuthMetadata.TokenUrl"/>, the spec-pinned endpoint.
    /// </param>
    /// <param name="skewSeconds">Refresh this many seconds before actual expiry.</param>
    /// <param name="httpTimeout">
    /// Token-endpoint per-call timeout override (tests inject a short one to exercise the
    /// timeout path without a 30s wait). Default: <see cref="TokenEndpointTimeout"/>.
    /// </param>
    public TokenManager(
        TokenStore store,
        ClientCredentials? credentials,
        Tokens? tokens,
        HttpMessageHandler? handler = null,
        string? tokenUrl = null,
        long skewSeconds = DefaultSkewSeconds,
        TimeSpan? httpTimeout = null)
    {
        _store = store;
        _credentials = credentials;
        _tokens = tokens;
        _tokenUrl = tokenUrl ?? OAuthMetadata.TokenUrl;
        _skewSeconds = skewSeconds;
        // The default transport REFUSES redirects (AllowAutoRedirect = false): a confidential
        // client must never re-POST the token form (client_id + client_secret + refresh_token)
        // to a 3xx Location host — .NET's HttpClient default WOULD follow it and leak the
        // secret. A test may inject its own handler for hermetic mock responses.
        _http = handler is null
            ? new HttpClient(new SocketsHttpHandler { AllowAutoRedirect = false })
            : new HttpClient(handler);
        _http.Timeout = httpTimeout ?? TokenEndpointTimeout;
    }

    /// <summary>
    /// Whether tokens are loaded (does not validate them, and does not imply a refresh is
    /// possible — refresh additionally needs the client-credentials record).
    /// </summary>
    public bool IsAuthenticated
    {
        get
        {
            _mutex.Wait();
            try
            {
                return _tokens is not null;
            }
            finally
            {
                _mutex.Release();
            }
        }
    }

    /// <summary>
    /// Return a valid access token, refreshing (and persisting the rotation) if it is
    /// expired or within the skew window.
    /// </summary>
    public async Task<string> GetAccessTokenAsync(CancellationToken cancellationToken = default)
    {
        await _mutex.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            var current = _tokens ?? throw new NotAuthenticatedException();
            if (current.IsExpired(_skewSeconds))
            {
                await RefreshCriticalSectionAsync(cancellationToken).ConfigureAwait(false);
            }
            return _tokens!.AccessToken;
        }
        finally
        {
            _mutex.Release();
        }
    }

    /// <summary>
    /// Force a refresh regardless of expiry (call this when the data plane returns 401),
    /// persisting the rotation. If another process already rotated, its fresher tokens are
    /// adopted instead of burning that rotation with a second endpoint call.
    /// </summary>
    public async Task ForceRefreshAsync(CancellationToken cancellationToken = default)
    {
        await _mutex.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            await RefreshCriticalSectionAsync(cancellationToken).ConfigureAwait(false);
        }
        finally
        {
            _mutex.Release();
        }
    }

    /// <summary>
    /// The reload → refresh → persist critical section, run under the store's exclusive
    /// lock so only one coordinated process rotates at a time. Caller holds
    /// <see cref="_mutex"/>.
    ///
    /// The adopt rule covers both entry points: if disk holds tokens that differ from
    /// memory and are not expired, another process already rotated — adopt them instead of
    /// re-burning the rotation. A refresh 400 (usually "our refresh token is stale") is
    /// retried ONCE against freshly reloaded disk state, which absorbs rotations by writers
    /// not honoring our lock — this protocol, not the lock's cross-runtime semantics, is
    /// the interop guarantee with the Rust CLI (see <see cref="TokenStore"/>).
    /// </summary>
    private async Task RefreshCriticalSectionAsync(CancellationToken cancellationToken)
    {
        var credentials = _credentials ?? throw new MissingClientCredentialsException();

        using var storeLock = await _store.AcquireLockAsync(cancellationToken).ConfigureAwait(false);

        if (_store.LoadTokens() is { } disk)
        {
            var differs = _tokens?.AccessToken != disk.AccessToken;
            if (differs && !disk.IsExpired(_skewSeconds))
            {
                _tokens = disk;
                return;
            }
            // Refresh from the freshest persisted rotation, never from stale memory.
            _tokens = disk;
        }
        var current = _tokens ?? throw new NotAuthenticatedException();

        Tokens refreshed;
        try
        {
            refreshed = await RefreshAtAsync(credentials, current, cancellationToken).ConfigureAwait(false);
        }
        catch (TokenEndpointException e) when (e.StatusCode == 400)
        {
            // If disk moved past what we sent (a rotation by an uncoordinated writer),
            // retry once with the fresher token before surfacing "re-login".
            if (_store.LoadTokens() is { } fresher && fresher.RefreshToken != current.RefreshToken)
            {
                refreshed = await RefreshAtAsync(credentials, fresher, cancellationToken).ConfigureAwait(false);
            }
            else
            {
                throw;
            }
        }
        _store.SaveTokens(refreshed);
        _tokens = refreshed;
    }

    private async Task<Tokens> RefreshAtAsync(
        ClientCredentials credentials,
        Tokens current,
        CancellationToken cancellationToken)
    {
        // Confidential client: the token endpoint requires client_id AND client_secret in
        // the form body (never in the URL — no secrets in query strings).
        using var content = new FormUrlEncodedContent(new Dictionary<string, string>
        {
            ["grant_type"] = "refresh_token",
            ["refresh_token"] = current.RefreshToken,
            ["client_id"] = credentials.ClientId,
            ["client_secret"] = credentials.ClientSecret,
        });

        HttpResponseMessage response;
        try
        {
            response = await _http.PostAsync(_tokenUrl, content, cancellationToken).ConfigureAwait(false);
        }
        catch (TaskCanceledException e) when (!cancellationToken.IsCancellationRequested)
        {
            // The hard token-endpoint timeout elapsed (it bounds lock-hold time). Surface a
            // typed transport error, not a bare TaskCanceledException. A caller-requested
            // cancellation is excluded by the filter and propagates as OperationCanceledException.
            throw new TransportException("token endpoint request timed out", e);
        }
        catch (HttpRequestException e)
        {
            throw new TransportException("token endpoint request failed", e);
        }

        using (response)
        {
            var status = (int)response.StatusCode;
            var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);

            // Defense-in-depth against a redirect leaking the confidential form: the default
            // transport already refuses to follow redirects, so a 3xx from the token endpoint
            // surfaces here rather than being followed. Reject it explicitly — a 2xx is the ONLY
            // success — so even an injected/misconfigured handler that leaves a bare 3xx in place
            // cannot slip past as a silent no-op.
            if (status is >= 300 and < 400)
            {
                throw new TokenEndpointException(status, body);
            }
            if (!response.IsSuccessStatusCode)
            {
                throw new TokenEndpointException(status, body);
            }

            // A hostile or broken 2xx body must fail as the typed TokenEndpointException, never a
            // raw JsonException detonating downstream and never a half-populated token persisted
            // (an empty access_token or a non-positive expiry would only resurface as a baffling
            // 400 on the NEXT refresh, long after the cause). These throws all run BEFORE the
            // store is written, so a hostile 2xx never burns the stored rotation. Messages are
            // FIXED and secret-free: the raw body is never echoed, since a partial 2xx payload may
            // carry token material. Mirrors go/auth/oauth.go:70-97.
            TokenResponse? parsed;
            try
            {
                parsed = JsonSerializer.Deserialize<TokenResponse>(body);
            }
            catch (JsonException)
            {
                throw new TokenEndpointException(status, "token-endpoint 2xx response was not valid JSON");
            }
            if (parsed is null)
            {
                // The body was the literal JSON null.
                throw new TokenEndpointException(status, "token-endpoint 2xx response was empty");
            }
            if (string.IsNullOrEmpty(parsed.AccessToken))
            {
                throw new TokenEndpointException(status, "token-endpoint 2xx response missing access_token");
            }
            if (parsed.ExpiresIn <= 0)
            {
                throw new TokenEndpointException(status, "token-endpoint 2xx response missing or invalid expires_in");
            }

            return new Tokens
            {
                AccessToken = parsed.AccessToken,
                // Persist the ROTATED refresh token; treat null OR empty like a missing one and
                // keep the current (still-valid) token. An empty refresh_token would clobber the
                // good one and 400 every future refresh.
                RefreshToken = string.IsNullOrEmpty(parsed.RefreshToken) ? current.RefreshToken : parsed.RefreshToken,
                ExpiresAt = DateTimeOffset.UtcNow.ToUnixTimeSeconds() + parsed.ExpiresIn,
                Scope = string.IsNullOrEmpty(parsed.Scope) ? current.Scope : parsed.Scope,
                TokenType = string.IsNullOrEmpty(parsed.TokenType) ? current.TokenType : parsed.TokenType,
            };
        }
    }

    /// <summary>Releases the token-endpoint HTTP client and the internal mutex.</summary>
    public void Dispose()
    {
        _http.Dispose();
        _mutex.Dispose();
    }

    /// <summary>Raw token-endpoint response shape.</summary>
    private sealed record TokenResponse
    {
        [JsonPropertyName("access_token")]
        public string? AccessToken { get; init; }

        [JsonPropertyName("refresh_token")]
        public string? RefreshToken { get; init; }

        [JsonPropertyName("expires_in")]
        public long ExpiresIn { get; init; }

        [JsonPropertyName("token_type")]
        public string? TokenType { get; init; }

        [JsonPropertyName("scope")]
        public string? Scope { get; init; }

        /// <summary>
        /// Redacts both token fields (parity with <see cref="Tokens"/> / <see cref="ClientCredentials"/>):
        /// the synthesized record ToString would otherwise print the raw access/refresh tokens if
        /// this value ever reached a log line.
        /// </summary>
        public override string ToString() =>
            "TokenResponse { access_token = [REDACTED], refresh_token = [REDACTED], " +
            $"expires_in = {ExpiresIn}, token_type = {TokenType ?? "null"}, scope = {Scope ?? "null"} }}";
    }
}

File: sdks/java/auth/src/main/java/com/ouratoolkit/auth/TokenManager.java
package com.ouratoolkit.auth;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import java.util.StringJoiner;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * The runtime auth layer: owns the current tokens, refreshes them proactively (expiry
 * skew), and persists every rotation. Mirrors the Rust companion's {@code TokenManager}
 * behavior so the two runtimes can safely share one on-disk store.
 *
 * <p><b>Refresh strategy:</b> proactive — {@link #getAccessToken()} refreshes when the
 * access token is expired or within the skew window, so requests carry a valid token.
 * Reactive refresh-on-401-then-retry is done by callers via {@link #forceRefresh()}.
 *
 * <p><b>Cross-process safety:</b> Oura invalidates the previous refresh token on every
 * rotation, and multiple processes (this SDK, the Rust CLI, the MCP server) share one
 * store. Every refresh therefore runs under the store's exclusive lock and RE-READS the
 * store first — if another process already rotated, its fresher tokens are ADOPTED
 * instead of burning (and thereby invalidating) that rotation with a second refresh. A
 * refresh that still 400s is retried ONCE against freshly reloaded disk state before
 * surfacing "re-login". Because Java's file lock does not interact with the Rust
 * runtime's {@code flock} on Linux (see {@link TokenStore}), this reload+retry protocol —
 * not the lock — is the cross-runtime guarantee.
 *
 * <p>The token endpoint has a hard {@link #DEFAULT_ENDPOINT_TIMEOUT 30s} timeout, which
 * bounds how long the store lock can be held (worst case ~2×: the 400-retry arm can chain
 * a second endpoint call under the same lock).
 */
public final class TokenManager {

    /** Refresh this many seconds before the token's actual expiry. */
    public static final long DEFAULT_SKEW_SECS = 60;

    /**
     * Hard timeout on each token-endpoint call — load-bearing: the call runs under the
     * store's exclusive lock, so an unbounded hang would wedge every other process
     * waiting on it.
     */
    public static final Duration DEFAULT_ENDPOINT_TIMEOUT = Duration.ofSeconds(30);

    private static final ObjectMapper MAPPER = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

    private final TokenStore store;
    private final ClientCredentials credentials; // nullable: tokens-only use is legal
    private final HttpClient http;
    private Tokens tokens; // guarded by `this`

    /** The spec-derived token endpoint; overridden only by tests (mock server). */
    private String tokenUrl = OuraOAuthMetadata.TOKEN_URL;

    private long skewSecs = DEFAULT_SKEW_SECS;
    private Duration endpointTimeout = DEFAULT_ENDPOINT_TIMEOUT;

    /**
     * Construct from an explicit store + optional in-memory records. Both records are
     * independently optional: credentials-without-tokens is "setup done, no login yet";
     * tokens-without-credentials is a caller-supplied token that can be used until expiry
     * but not refreshed ({@link MissingClientCredentialsException}).
     */
    public TokenManager(TokenStore store, ClientCredentials credentials, Tokens tokens) {
        this.store = store;
        this.credentials = credentials;
        this.tokens = tokens;
        this.http = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                // Pin the JDK default: NEVER follow redirects from the token endpoint. A
                // 3xx from a hostile/misconfigured endpoint must NOT silently re-POST the
                // confidential-client form (client_id + client_secret + refresh_token) to
                // the redirect target — it surfaces as a typed non-2xx error instead.
                .followRedirects(HttpClient.Redirect.NEVER)
                .build();
    }

    /**
     * Load from the default per-platform store. Absent records are not an error —
     * {@link #getAccessToken()} reports {@link NotAuthenticatedException} on first use.
     */
    public static TokenManager load() throws AuthException {
        TokenStore store = TokenStore.openDefault();
        try {
            return new TokenManager(
                    store,
                    store.loadCredentials().orElse(null),
                    store.loadTokens().orElse(null));
        } catch (IOException e) {
            throw new StoreException("reading the token store failed", e);
        }
    }

    /** Whether tokens are loaded (does not validate them or imply refresh is possible). */
    public synchronized boolean isAuthenticated() {
        return tokens != null;
    }

    /**
     * Return a valid access token, refreshing (and persisting the rotation) if it is
     * expired or within the skew window.
     */
    public synchronized String getAccessToken() throws AuthException {
        if (tokens == null) {
            throw new NotAuthenticatedException();
        }
        if (tokens.isExpired(skewSecs)) {
            refreshCriticalSection();
        }
        return tokens.getAccessToken();
    }

    /**
     * Force a refresh regardless of expiry (used by callers on a 401), persisting the
     * rotation. If another process already rotated, its fresher tokens are adopted
     * instead of burning that rotation with a second refresh.
     */
    public synchronized void forceRefresh() throws AuthException {
        refreshCriticalSection();
    }

    /**
     * The reload → refresh → persist critical section, run under the store's exclusive
     * lock. The adopt rule covers both entry points: if disk holds tokens that differ
     * from memory and aren't expired, another process already rotated — adopt them. (On
     * the proactive path memory is expired, so anything fresher is strictly better; on
     * the force path memory just 401'd, so a DIFFERENT fresh token is the fix and an
     * identical one means we must rotate.)
     */
    private void refreshCriticalSection() throws AuthException {
        if (credentials == null) {
            throw new MissingClientCredentialsException();
        }
        try (TokenStore.StoreLock ignored = lockStore()) {
            Optional<Tokens> diskOpt = loadTokensFromDisk();
            if (diskOpt.isPresent()) {
                Tokens disk = diskOpt.get();
                boolean differs = tokens == null
                        || !tokens.getAccessToken().equals(disk.getAccessToken());
                if (differs && !disk.isExpired(skewSecs)) {
                    tokens = disk; // adopt the rotation another process performed
                    return;
                }
                // Refresh from the freshest persisted rotation, never from stale memory.
                tokens = disk;
            }
            if (tokens == null) {
                throw new NotAuthenticatedException();
            }

            Tokens refreshed;
            try {
                refreshed = callTokenEndpoint(tokens);
            } catch (TokenEndpointException e) {
                // A 400 usually means the refresh token we sent is no longer valid. If
                // disk moved past what we sent (a rotation by a writer not coordinated
                // through this lock — e.g. the Rust runtime on Linux), retry ONCE with
                // the fresher token before surfacing "re-login".
                if (e.getStatus() != 400) {
                    throw e;
                }
                Optional<Tokens> reloaded = loadTokensFromDisk();
                if (reloaded.isPresent()
                        && !reloaded.get().getRefreshToken().equals(tokens.getRefreshToken())) {
                    tokens = reloaded.get();
                    refreshed = callTokenEndpoint(tokens);
                } else {
                    throw e;
                }
            }

            try {
                store.saveTokens(refreshed);
            } catch (IOException e) {
                throw new StoreException("persisting the rotated tokens failed", e);
            }
            tokens = refreshed;
        }
    }

    private TokenStore.StoreLock lockStore() throws StoreException {
        try {
            return store.lockExclusive();
        } catch (IOException e) {
            throw new StoreException("locking the token store failed", e);
        }
    }

    private Optional<Tokens> loadTokensFromDisk() throws StoreException {
        try {
            return store.loadTokens();
        } catch (IOException e) {
            throw new StoreException("reading the token store failed", e);
        }
    }

    /**
     * POST the refresh grant (confidential client: {@code client_id} + {@code
     * client_secret} in the form body — never in a URL query string). Oura ROTATES the
     * refresh token: the returned {@link Tokens} carry the new one, falling back to the
     * current one only if the server omits it.
     */
    private Tokens callTokenEndpoint(Tokens current) throws AuthException {
        Map<String, String> form = new LinkedHashMap<>();
        form.put("grant_type", "refresh_token");
        form.put("refresh_token", current.getRefreshToken());
        form.put("client_id", credentials.getClientId());
        form.put("client_secret", credentials.getClientSecret());

        HttpRequest request = HttpRequest.newBuilder(URI.create(tokenUrl))
                .timeout(endpointTimeout)
                .header("Content-Type", "application/x-www-form-urlencoded")
                .POST(HttpRequest.BodyPublishers.ofString(encodeForm(form)))
                .build();

        final HttpResponse<String> response;
        try {
            response = http.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (IOException e) {
            // Includes HttpTimeoutException: the hard timeout that bounds lock-hold time.
            throw new TransportException("token endpoint request failed", e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new TransportException("interrupted during token refresh", e);
        }

        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new TokenEndpointException(response.statusCode(), response.body());
        }

        final JsonNode node;
        try {
            node = MAPPER.readTree(response.body());
        } catch (IOException e) {
            throw new TransportException("token endpoint returned unparseable JSON", e);
        }
        // A hostile or broken 2xx body must fail as a typed error, never a half-populated
        // token persisted to the store (an empty access_token or a non-positive expiry
        // would only resurface as a baffling 400 on the NEXT refresh, long after the cause).
        // Mirrors go/auth/oauth.go:74-79. Messages carry NO token/secret material — the raw
        // body is never echoed, since a partial 2xx payload may contain token material.
        JsonNode accessToken = node == null ? null : node.get("access_token");
        JsonNode expiresIn = node == null ? null : node.get("expires_in");
        if (accessToken == null || !accessToken.isTextual() || accessToken.asText().isEmpty()) {
            throw new TransportException(
                    "token endpoint 2xx response missing or empty access_token",
                    null);
        }
        // Reject 0, negative, AND non-numeric (e.g. "expires_in":"abc", where a bare
        // asLong() would silently coerce to 0 and be treated as immediately expired).
        if (expiresIn == null || !expiresIn.canConvertToLong() || expiresIn.asLong() <= 0) {
            throw new TransportException(
                    "token endpoint 2xx response missing or invalid expires_in",
                    null);
        }
        String rotatedRefresh = node.hasNonNull("refresh_token")
                ? node.get("refresh_token").asText()
                : current.getRefreshToken(); // server omitted rotation; keep the old one
        String scope = node.hasNonNull("scope") ? node.get("scope").asText() : current.getScope();
        String tokenType = node.hasNonNull("token_type")
                ? node.get("token_type").asText()
                : current.getTokenType();
        return new Tokens(
                accessToken.asText(),
                rotatedRefresh,
                Instant.now().getEpochSecond() + expiresIn.asLong(),
                scope,
                tokenType);
    }

    private static String encodeForm(Map<String, String> form) {
        StringJoiner joiner = new StringJoiner("&");
        for (Map.Entry<String, String> e : form.entrySet()) {
            joiner.add(URLEncoder.encode(e.getKey(), StandardCharsets.UTF_8)
                    + "="
                    + URLEncoder.encode(e.getValue(), StandardCharsets.UTF_8));
        }
        return joiner.toString();
    }

    // --- test seams (package-private; hermetic tests point at a loopback endpoint) ------

    void overrideTokenUrl(String url) {
        this.tokenUrl = url;
    }

    void overrideSkewSecs(long skewSecs) {
        this.skewSecs = skewSecs;
    }

    void overrideEndpointTimeout(Duration timeout) {
        this.endpointTimeout = timeout;
    }

    /** Never leaks tokens or the client secret. */
    @Override
    public String toString() {
        return "TokenManager{store=" + store.getDir()
                + ", credentials=" + credentials
                + ", tokens=" + tokens + "}";
    }
}
